### PR TITLE
print an error message when a non-git gem is given a `branch` option

### PR DESCRIPTION
**Pull request by colby-swandale**
When the user supplies the `branch` option to a non-git gem an error message should be printed.

Relates to #5530.

### CODE REVIEW

File: lib/bundler/dsl.rb [hunk 1]
```diff
@@ -321,6 +321,10 @@ def normalize_options(name, version, opts)
 
       normalize_hash(opts)
 
+      if opts["branch"] && !(opts["git"] || opts["github"])
```
Comment (segiddins) [inline lib/bundler/dsl.rb]: Why not do this in `validate_keys`?

Comment (colby-swandale) [inline lib/bundler/dsl.rb]: I was being lazy and really should have checked. Thanks for the feedback!

File: spec/bundler/dsl_spec.rb [hunk 2]
```diff
@@ -144,6 +144,11 @@
       expect { subject.gem(:foo) }.
         to raise_error(Bundler::GemfileError, /You need to specify gem names as Strings. Use 'gem "foo"' instead/)
     end
+
+    it "rejects branch option on non-git gems" do
+      expect { subject.gem("foo", :branch => "test") }.
+        to raise_error(Bundler::GemfileError, /The `branch` option for `gem "foo"` is not allowed. Only gems with a git source can specify a branch/)
```
Comment (segiddins) [inline spec/bundler/dsl_spec.rb]: quotes don't match, hence the test failure

File: spec/bundler/dsl_spec.rb [hunk 3]
```diff
@@ -144,6 +144,11 @@
       expect { subject.gem(:foo) }.
         to raise_error(Bundler::GemfileError, /You need to specify gem names as Strings. Use 'gem "foo"' instead/)
     end
+
+    it "rejects branch option on non-git gems" do
```
Comment (segiddins) [inline spec/bundler/dsl_spec.rb]: please also test that specifying a branch when using a user-defined git source still passes? (or make sure such a test exists?)

**Comment by segiddins** [comment]
@homu r+

**Comment by segiddins** [comment]
@bundlerbot r+

**Pull request merged**